trio/_core/_traps.py

```diff
@@ -163,3 +164,107 @@ def abort(inner_raise_cancel):
 
     """
     return (await _async_yield(WaitTaskRescheduled(abort_func))).unwrap()
+
+
+# Not exported in the trio._core namespace, but imported directly by _run.
+@attr.s(frozen=True)
+class PermanentlyDetachCoroutineObject:
+    final_outcome = attr.ib()
+
+
+async def permanently_detach_coroutine_object(final_outcome):
+    """Permanently detach the current task from the Trio scheduler.
+
+    Normally, a Trio task doesn't exit until its coroutine object exits. When
+    you call this function, Trio acts like the coroutine object just exited
+    and the task terminates with the given outcome. This is useful if you want
+    to permanently switch the coroutine object over to a different coroutine
+    runner.
+
+    When the calling coroutine enters this function it's running under Trio,
+    and when the function returns it's running under the foreign coroutine
+    runner.
+
+    You should make sure that the coroutine object has released any
+    Trio-specific resources it has acquired (e.g. nurseries).
+
+    Args:
+      final_outcome (outcome.Outcome): Trio acts as if the current task exited
+          with the given return value or exception.
+
+    Returns or raises whatever value or exception the new coroutine runner
+    uses to resume the coroutine.
+
+    """
+    if _run.current_task().child_nurseries:
```

Comment:
Should this also check whether the detaching task is in any cancel scopes? I think task_exited will remove it from the scopes' task lists, but exiting the cancel scope in the now-detached task will then create some sort of confusing error.

Reply:
I thought about that, but we don't really have an easy way to check if the detached task has any `with <cancel scope>` blocks open. We can see whether there are any cancel scopes active, but that includes inherited scopes, that don't correspond to a `with` block inside the exiting task. And, if someone does detach inside a `with <cancel scope>` block, then it's true that they'll eventually get some kind of explosion in the `__exit__` method, but I don't think this will corrupt Trio's internal state; it'll just be a regular exception.

For nurseries OTOH, it's easy to check for, and killing a task with active nurseries inside would break Trio's internal invariants.

Of course there are lots of ways to misuse `trio.hazmat` and these functions in particular to break Trio's internal invariants, but since the check was easy I figured why not.

Reply:
Makes sense! I hadn't checked in detail as to whether the cancel scope check would be feasible, and my memory of the bookkeeping in CancelScope was incorrect.